[Doc] Add DeepCache in section `optimization/General optimizations` (#6390)
huggingface/diffusers: docs/source/en/optimization/deepcache.md

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# DeepCache

> [DeepCache](https://huggingface.co/papers/2312.00858) accelerates [`StableDiffusionPipeline`] by strategically caching and reusing high-level features, while efficiently updating low-level features, which leverages the unique properties of the U-Net architecture. 

Reviewer: If it supports `StableDiffusionXLPipeline` too, I'd mention it here. 

Additionally, out of curiosity: does it support the img2img, and inpainting pipelines from SD and SDXL yet? If not, it might be extremely beneficial for the community to have them supported because they are so widely used.

Author: Yes, it supports StableDiffusionXLPipeline!

For other tasks, like inpainting, img2img, and video, I listed some of the results here:

| Task | Model | Original Pipeline | + DeepCache | Speedup |
| ---- | ------ | ----------------|-------------| ------- |
| inpainting | diffusers/stable-diffusion-xl-1.0-inpainting-0.1 | <img width="400" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/28f02afc-a060-4a1f-8205-86a47c2f939c"> |<img width="400" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/168ce177-06bc-4633-9849-d2010e739029"> | 1.8x |
| inpainting | runwayml/stable-diffusion-inpainting | <img width="467" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/c003b73f-fce2-4ed1-b383-3d621d72b597"> |  <img width="468" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/523dbf14-4fc9-412b-97e1-b2ce2d78276f"> | 1.7x |
| img2img | runwayml/stable-diffusion-v1-5 | <img width="400" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/2afb36e1-b6f7-49f7-8139-ad0b14857553"> | <img width="390" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/59afd43f-20d2-4a75-9b21-891def9874a5"> | 1.6x |
| video | stabilityai/stable-video-diffusion-img2vid-xt | ![origin](https://github.com/huggingface/diffusers/assets/22924514/c98978ed-4d4f-4d82-8dfa-3614ce62aaf1) | ![deepcache](https://github.com/huggingface/diffusers/assets/22924514/ccdd42fa-98ef-48af-9b07-5404461eabb3) | 1.7x |

In these cases, under the premise of ensuring quality and consistency, the acceleration ratio obtained is not particularly high. 

I would later add scripts of these cases in the DeepCache repo😃

Install DeepCache from `pip`:
```bash
pip install DeepCache
```

You can use [`DeepCache`](https://github.com/horseee/DeepCache) by loading and enabling the [`DeepCacheSDHelper`](https://github.com/horseee/DeepCache#usage):

```diff
  import torch
  from diffusers import StableDiffusionPipeline
  pipe = StableDiffusionPipeline.from_pretrained('runwayml/stable-diffusion-v1-5', torch_dtype=torch.float16).to("cuda")

+ from DeepCache import DeepCacheSDHelper
+ helper = DeepCacheSDHelper(pipe=pipe)
+ helper.set_params(
+     cache_interval=3,
+     cache_branch_id=0,
+ )
+ helper.enable()

  image = pipe("a photo of an astronaut on a moon").images[0]
```

The `set_params` method accepts two arguments: `cache_interval` and `cache_branch_id`. `cache_interval` means the frequency of feature caching, specified as the number of steps between each cache operation. `cache_branch_id` identifies which branch of the network (ordered from the shallowest to the deepest layer) is responsible for executing the caching processes. 

> Opting for a lower `cache_branch_id` or a larger `cache_interval` can lead to faster inference speed; however, this may come at the cost of reduced image quality. Once those arguments are set, use the `enable` or `disable` methods to activate or deactivate the DeepCacheSDHelper respectively.

Reviewer: Very clear! If there's any ablation already done on these hyperparameters in the original DeepCache paper, I'd mention it here too.

Author: The ablation experiments for `cache_interval` are detailed in Tables 13 and 14 of the appendix, using SD1.5 as the base model. Regarding `cache_branch_id`, its corresponding ablation experiments are also presented, found in Table 12; however, these experiments were carried out using the DDPM model. 

* cache_interval: (N here represents the cache interval)
<img width="501" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/e51c1d33-5243-45d6-a696-0c794c75ce60">

* cache_branch_id:
<img width="299" alt="image" src="https://github.com/huggingface/diffusers/assets/22924514/0edf96ce-a51f-4fd2-a868-6fcf6c5d7c23">

Would this be okay?

Reviewer: Sure. Feel free to include this in the doc.

<div class="flex justify-center">
    <img src="https://github.com/horseee/Diffusion_DeepCache/raw/master/static/images/example.png">
</div>

> You can find more generated samples (Original Pipeline v.s. DeepCache) and the corresponding inference latency in the [WanDB report](https://wandb.ai/horseee/DeepCache/runs/jwlsqqgt?workspace=user-horseee). The prompts are randomly selected from the [MS-COCO 2017](https://cocodataset.org/#home) dataset.

Reviewer: ```suggestion
You can find more generated samples (Original Pipeline v.s. DeepCache) and the corresponding inference latency in the [WandB report](https://wandb.ai/horseee/DeepCache/runs/jwlsqqgt?workspace=user-horseee). The prompts are randomly selected from the [MS-COCO 2017](https://cocodataset.org/#home) dataset.
```

## Benchmark

> We measure the acceleration ratio achievable using DeepCache. All evaluations are based on the [Stable Diffusion v2.1](https://huggingface.co/stabilityai/stable-diffusion-2-1) with 50 inference steps, using NVIDIA RTX A5000. The results show the speed enhancements that can be expected under different configurations of resolution, batch size, the interval for cache(I) and the branch for cache(B).

Reviewer 2: ```suggestion
We tested how much faster DeepCache accelerates [Stable Diffusion v2.1](https://huggingface.co/stabilityai/stable-diffusion-2-1) with 50 inference steps on an NVIDIA RTX A5000, using different configurations for resolution, batch size, cache interval (I), and cache branch (B).
```
